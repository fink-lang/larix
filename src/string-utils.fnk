unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
