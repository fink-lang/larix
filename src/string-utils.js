import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

