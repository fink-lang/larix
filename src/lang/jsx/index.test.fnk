{describe, it, eq, to_throw, snapshot} = import '../../testing/jest'

{parse_expr} = import '../../'


describe:: 'JSX <Foobar>...</Foobar>', fn:

  it:: 'parses self closing elem: <Foobar />', fn:
    parse_expr(`<Foobar />`) eq snapshot


  it:: 'parses self closing elem with short attr: <Foobar spam />', fn:
    parse_expr(`<Foobar spam />`) eq snapshot


  it:: 'parses self closing elem with str attr: <Foobar spam="ni" />', fn:
    parse_expr(`<Foobar spam="ni" />`) eq snapshot


  it:: 'parses self closing elem with str attr: <Foobar spam=\'ni\' />', fn:
    parse_expr(`<Foobar spam='ni' />`) eq snapshot


  it:: 'parses self closing elem with expr attr: <Foobar spam={ni} />', fn:
    parse_expr(`<Foobar spam={ni} />`) eq snapshot


  it:: 'parses empty elem: <Foobar></Foobar>', fn:
    parse_expr(`<Foobar></Foobar>`) eq snapshot


  it:: 'parses elem with children: <Foobar><Spam /></Foobar>', fn:
    parse_expr(`
      <Foobar ni>
        <Spam />
      </Foobar>
    `) eq snapshot


  it:: 'parses elem with expr in body: <Foobar> {...} </Foobar>', fn:
    parse_expr(`
      <Foobar>
        {1 + 2}
      </Foobar>
    `) eq snapshot


describe:: 'JSX parse errors', fn:
  it:: 'throws with invalid attr value', fn:

    (
      fn: parse_expr:: '<Foobar spam=123 />', 'test.fnk'
    ) to_throw `
      test.fnk:1:13
      1| <Foobar spam=123 />
                      ^

      Expected '{' but found '123'.
    `
