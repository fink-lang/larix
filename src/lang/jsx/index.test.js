import {strip_block} from '../../string-utils';
import {parse_expr} from '../../';


describe('JSX <Foobar>...</Foobar>', ()=> {

  it('parses self closing elem: <Foobar />', ()=> {
    expect(
      parse_expr(`<Foobar />`)
    ).toMatchSnapshot();
  });


  it('parses self closing elem with short attr: <Foobar spam />', ()=> {
    expect(
      parse_expr(`<Foobar spam />`)
    ).toMatchSnapshot();
  });


  it('parses self closing elem with str attr: <Foobar spam="ni" />', ()=> {
    expect(
      parse_expr(`<Foobar spam="ni" />`)
    ).toMatchSnapshot();
  });


  it('parses self closing elem with str attr: <Foobar spam=\'ni\' />', ()=> {
    expect(
      parse_expr(`<Foobar spam='ni' />`)
    ).toMatchSnapshot();
  });


  it('parses self closing elem with expr attr: <Foobar spam={ni} />', ()=> {
    expect(
      parse_expr(`<Foobar spam={ni} />`)
    ).toMatchSnapshot();
  });


  it('parses empty elem: <Foobar></Foobar>', ()=> {
    expect(
      parse_expr(`<Foobar></Foobar>`)
    ).toMatchSnapshot();
  });


  it('parses elem with children: <Foobar><Spam /></Foobar>', ()=> {
    expect(
      parse_expr(strip_block`
        <Foobar ni>
          <Spam />
        </Foobar>
      `)
    ).toMatchSnapshot();
  });


  it('parses elem with expr in body: <Foobar> {...} </Foobar>', ()=> {
    expect(
      parse_expr(strip_block`
        <Foobar>
          {1 + 2}
        </Foobar>
      `)
    ).toMatchSnapshot();
  });
});


describe('JSX parse errors', ()=> {
  it('throws with invalid attr value', ()=> {
    expect(()=> parse_expr(
      '<Foobar spam=123 />', 'test.fnk'
    )).toThrow(strip_block`
      test.fnk:1:13
      1| <Foobar spam=123 />
                      ^

      Expected '{' but found '123'.
    `);
  });
});

