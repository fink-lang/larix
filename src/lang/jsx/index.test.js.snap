// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX <Foobar>...</Foobar> parses elem with children: <Foobar><Spam /></Foobar> 1`] = `
jsx-elem (1:0-3:9) Foobar
  jsx-attr (1:8-1:10) ni
    null
  :
  jsx-text (1:10-1:11)
    "\\n  "
  jsx-elem (2:2-2:10) Spam
  jsx-text (2:8-2:10)
    "\\n"
`;

exports[`JSX <Foobar>...</Foobar> parses elem with expr in body: <Foobar> {...} </Foobar> 1`] = `
jsx-elem (1:0-3:9) Foobar
  :
  jsx-text (1:7-1:8)
    "\\n  "
  jsx-expr-container (2:2-2:8)
    block (2:2-2:8)
      arithm + (2:3-2:8)
        number (2:3-2:4) 1
        number (2:7-2:8) 2
  jsx-text (2:8-2:9)
    "\\n"
`;

exports[`JSX <Foobar>...</Foobar> parses empty elem: <Foobar></Foobar> 1`] = `jsx-elem (1:0-1:17) Foobar`;

exports[`JSX <Foobar>...</Foobar> parses self closing elem with expr attr: <Foobar spam={ni} /> 1`] = `
jsx-elem (1:0-1:20) Foobar
  jsx-attr (1:8-1:17) spam
        jsx-expr-container (1:13-1:16)
      block (1:13-1:16)
        ident (1:14-1:16) ni
`;

exports[`JSX <Foobar>...</Foobar> parses self closing elem with short attr: <Foobar spam /> 1`] = `
jsx-elem (1:0-1:15) Foobar
  jsx-attr (1:8-1:12) spam
    null
`;

exports[`JSX <Foobar>...</Foobar> parses self closing elem with str attr: <Foobar spam="ni" /> 1`] = `
jsx-elem (1:0-1:20) Foobar
  jsx-attr (1:8-1:17) spam
        jsx-string (1:13-1:17)
      'ni'
`;

exports[`JSX <Foobar>...</Foobar> parses self closing elem with str attr: <Foobar spam='ni' /> 1`] = `
jsx-elem (1:0-1:20) Foobar
  jsx-attr (1:8-1:17) spam
        jsx-string (1:13-1:17)
      'ni'
`;

exports[`JSX <Foobar>...</Foobar> parses self closing elem: <Foobar /> 1`] = `jsx-elem (1:0-1:10) Foobar`;
